xds/src/main/java/io/grpc/xds/RingHashLoadBalancer.java

```diff
@@ -0,0 +1,455 @@
+  }
+
+  private static EquivalentAddressGroup stripAttrs(EquivalentAddressGroup eag) {
+    return new EquivalentAddressGroup(eag.getAddresses());
+  }
+
+  private static Ref<ConnectivityStateInfo> getSubchannelStateInfoRef(
+      Subchannel subchannel) {
+    return checkNotNull(subchannel.getAttributes().get(STATE_INFO), "STATE_INFO");
+  }
+
+  private static final class RingHashPicker extends SubchannelPicker {
+    private final SynchronizationContext syncContext;
+    private final List<RingEntry> ring;
+    // Avoid synchronization between pickSubchannel and subchannel's connectivity state change,
+    // freeze picker's view of subchannel's connectivity state.
+    private final Map<EquivalentAddressGroup, SubchannelView> pickableSubchannels;  // read-only
+
+    private RingHashPicker(
+        SynchronizationContext syncContext, List<RingEntry> ring,
+        Map<EquivalentAddressGroup, Subchannel> subchannels) {
+      this.syncContext = syncContext;
+      this.ring = ring;
+      pickableSubchannels = new HashMap<>(subchannels.size());
+      for (Map.Entry<EquivalentAddressGroup, Subchannel> entry : subchannels.entrySet()) {
+        Subchannel subchannel = entry.getValue();
+        ConnectivityStateInfo stateInfo = subchannel.getAttributes().get(STATE_INFO).value;
+        pickableSubchannels.put(entry.getKey(), new SubchannelView(subchannel, stateInfo));
+      }
+    }
+
+    @Override
+    public PickResult pickSubchannel(PickSubchannelArgs args) {
+      Long requestHash = args.getCallOptions().getOption(XdsNameResolver.RPC_HASH_KEY);
+      if (requestHash == null) {
+        return PickResult.withError(RPC_HASH_NOT_FOUND);
+      }
+      if (ring.isEmpty()) {
+        return PickResult.withNoResult();
+      }
+
+      // Find the ring entry with hash next to (clockwise) the RPC's hash.
+      int low = 0;
+      int high = ring.size();
+      int mid;
+      while (true) {
+        mid = (low + high) / 2;
+        if (mid == ring.size()) {
+          mid = 0;
+          break;
+        }
+        long midVal = ring.get(mid).hash;
+        long midValL = mid == 0 ? 0 : ring.get(mid - 1).hash;
+        if (requestHash <= midVal && requestHash > midValL) {
+          break;
+        }
+        if (midVal < requestHash) {
+          low = mid + 1;
+        } else {
+          high =  mid - 1;
+        }
+        if (low > high) {
+          mid = 0;
+          break;
+        }
+      }
+
+      // Try finding a READY subchannel. Starting from the ring entry next to the RPC's hash.
+      // If the one of the first two subchannels is not in TRANSIENT_FAILURE, return result
+      // based on that subchannel. Otherwise, fail the pick unless a READY subchannel is found.
+      // Meanwhile, trigger connection for the first subchannel that is in IDLE if no subchannel
+      // before it is in CONNECTING or READY.
+      boolean hasPending = false;
+      Set<EquivalentAddressGroup> attemptedAddrs = new HashSet<>();
```

Comment:
This seems rather expensive for us to maintain each RPC pick. Why bother? It seems that we will just be looking at the SubchannelView's state, so just doing `contains(addrKey)` is the same cost as what it tries to avoid (because of `pickableSubchannels.get(addrKey)`. The `attemptedAddrs.size() <= 2` could just be `i < 2` and be generally the same thing.

Reply:
Well, this is actually necessary. Note we are iterating through the ring entries, and each address could be mapped to a lot of ring entries (that's how this algorithm supposed to work). Say we have a big ring (the bigger the ring, the closer RPCs distributed regarding to the desired proportion), 3000 entries, for 3 backends (for simplicity, same weight). Then 1000 entries are assigned for each backend, and entries for each backend are mixed together. Something like `[host0, host0, host2, host1, host1, host0, host1, ...]`.

Since we are counting against how many subchannels/hosts attempted, _instead of how many ring entries_. We need to bookkeep how many unique subchannels/addresses we attempted while iterating through the ring.

Reply:
This was not part of the design. If it is necessary the design needs to be updated and we make sure that the other languages do something similar.

The only "counting" that matters is for the very first two entries we check. If we want to make sure they aren't a duplicate all we need is `EquivalentAddressGroup firstAddr` and `boolean seenSecondAddr` variables , not a full set of everything encountered thus far.

But again, the design didn't call for that; it is either not a problem and so we shouldn't worry about it or it is a problem and everyone should worry about it.

Reply:
Looknig at the design, I can see the trouble:

> Move on to the next subchannel in the ring (i.e., the one it would have chosen based on the request's hash if the subchannel that is in state TRANSIENT_FAILURE had not been present in the ring).

It is unclear whether the parenthetical meant removing the ring entry or all entries for the subchannel. Taken alone I'd agree it would be all entries for the subchannel, but at that point it disagrees with the thing it is explaining. So I read it as imprecise language without thinking about it. If the parenthetical said "if the entry containing the subchannel that is in state TRANSIENT_FAILURE" (which is a mouth-full), then it'd be clearly "only one entry."

But even if that section is unclear, there pseudocode is very clear.

Reply:
But then the algorithm doesn't make sense. For large rings (each address being mapped to a lot of ring entries), it's highly likely you will just end up making decisions with the same subchannel referenced by different ring entries. At the first moment I saw the pseudocode, I assumed that was just high-level idea, `ring[first_index].subchannel` and `ring[second_index].subchannel` should really mean two different subchannels. That is, the algorithm is based on subchannels although it is walking through the ring.

@markdroth is that what you mean?

Reply:
Slightly optimized to eliminate the use of Set for bookkeeping subchannels have been checked.

Further optimization can be done to avoid usages of EAG as map keys (put a TODO there).